### PR TITLE
add source tags to waypoint trace

### PR DESCRIPTION
**Pull request by PetrMc**
**Please provide a description of this PR:**

- add tracing for source workloads. depends on https://github.com/istio/proxy/pull/6765 
- associated issue https://github.com/istio/istio/issues/58348

### CODE REVIEW

File: pilot/pkg/networking/core/listener_waypoint.go [hunk 1]
```diff
@@ -853,6 +853,11 @@ func (lb *ListenerBuilder) buildWaypointHTTPFilters(svc *model.Service) (pre []*
 	)
 	// TODO: how to deal with ext-authz? It will be in the ordering twice
 	// TODO policies here will need to be different per-chain (service attached)
+
+	// Add downstream peer metadata filter to main_internal chains so source workload
+	// metadata is available for tracing in this HCM (not just in connect_terminate)
+	pre = append(pre, xdsfilters.WaypointDownstreamMetadataFilter)
```
Comment (keithmattix) [inline pilot/pkg/networking/core/listener_waypoint.go:859]: I _think_ we store the filter state in a manner that ensures it's shared with upstream, so the addition of this filter may not be necessary (/cc @krinkinmu to keep me honest)

Comment (PetrMc) [reply pilot/pkg/networking/core/listener_waypoint.go:859]: thanks @keithmattix 

- `shared_with_upstream` in `peer_metadata` config controls sharing filter state to upstream connections. 
- `main_internal` HCM is a separate filter chain on the same listener - it needs its own `peer_metadata` filter to access the downstream peer info for tracing tags. 

Without it, the `FIELD accessor` format strings return empty because there's no `WorkloadMetadataObject` in that HCM's filter state.

it was tested end-to-end: without this filter addition, source tags are empty. With it, they populate correctly.

Comment (keithmattix) [reply pilot/pkg/networking/core/listener_waypoint.go:859]: Ah for upstream filters I imagine? Makes sense; thanks for clarifying

Comment (keithmattix) [reply pilot/pkg/networking/core/listener_waypoint.go:859]: I was doing some other proxy work and came back around to think about this: if this is true, why does peer_metadata live on connect_terminate in the first place? Does istio_stats not read anything from the downstream peer in waypoints today? That seems odd

Comment (PetrMc) [reply pilot/pkg/networking/core/listener_waypoint.go:859]: This is a valid point Keith. Let me dig more and will let you know (hopefully tomorrow) if it's my tests that somehow were returning the FIELD accessor tags empty.

Comment (PetrMc) [reply pilot/pkg/networking/core/listener_waypoint.go:859]: @keithmattix - thanks for flagging it again! The follow-up PR is https://github.com/istio/istio/pull/58866

it also includes suggestions from @Stevenjin8 - https://github.com/istio/istio/pull/58851#discussion_r2712877979
